src/beamlime/applications/base.py

```diff
@@ -0,0 +1,176 @@
+            )
+        self._break_routing_loop = True
+
+    @contextmanager
+    def handler_wrapper(
+        self, handler: Callable[..., Any], message: BeamlimeMessage
+    ) -> Generator[None, None, None]:
+        import warnings
+
+        try:
+            self.debug(f"Routing event {type(message)} to handler {handler}...")
+            yield
+        except Exception as e:
+            warnings.warn(f"Failed to handle event {type(message)}: {e}", stacklevel=2)
+        else:
+            self.debug(f"Routing event {type(message)} to handler {handler} done.")
+
+    def register_awaitable_handler(
+        self, event_tp, handler: Callable[[BeamlimeMessage], Coroutine[Any, Any, Any]]
+    ):
+        if event_tp in self.awaitable_handlers:
+            self.awaitable_handlers[event_tp].append(handler)
+        else:
+            self.awaitable_handlers[event_tp] = [handler]
+
+    def register_handler(self, event_tp, handler: Callable[[BeamlimeMessage], Any]):
+        if event_tp in self.handlers:
+            self.handlers[event_tp].append(handler)
+        else:
+            self.handlers[event_tp] = [handler]
+
+    async def route(self, message: BeamlimeMessage) -> None:
+        import warnings
+
+        if handlers := self.handlers.get(type(message), []):
+            for handler in handlers:
+                await asyncio.sleep(0)
+                with self.handler_wrapper(handler, message):
+                    handler(message)
+        if awaitable_handlers := self.awaitable_handlers.get(type(message), []):
+            for handler in awaitable_handlers:
+                with self.handler_wrapper(handler, message):
+                    await handler(message)
+        if not (handlers or awaitable_handlers):
+            warnings.warn(
+                f"No handler for event {type(message)}. Ignoring...", stacklevel=2
+            )
+
+    async def run(self) -> None:
+        data_monitor = self.data_pipe_monitor(
+            self.message_pipe,
+            timeout=0.1,
+            interval=0.1,
+        )
+
+        while not self._break_routing_loop:
+            if await data_monitor():
+                message = self.message_pipe.pop(0)
+                await self.route(message)
+
+        self.debug("Breaking routing loop. Routing the rest of the message...")
+        for message in self.message_pipe:
+            await self.route(message)
+        self.debug("Routing the rest of the message done.")
+
+    async def send_message_async(self, message: BeamlimeMessage) -> None:
+        self.message_pipe.append(message)
+        await asyncio.sleep(0)
+
+
+class BaseHandler(LogMixin, ABC):
+    """Base class for message handlers.
+
+    Message handlers are expected to register methods to a message router,
+    so that the message router can trigger the methods
+    with the certain type of message.
+
+    Registered methods should accept a message as the only positional argument.
```

Comment:
This seems to be a new mechanism? Not sure why this was added now? I suppose it has not been benchmarked? Is this required, given the quite static nature of Beamlime, which will always have the same components?

Reply:
But it's basically doing the same thing, and makes much less coroutine calls,
because in the original one, all daemons had to monitor their own lists,
but now it's only one instance that monitors the queue...